src/Workspaces/Core/Portable/Workspace/Solution/ProjectDependencyGraph.cs

```diff
@@ -14,35 +14,288 @@ namespace Microsoft.CodeAnalysis
+                  reverseTransitiveReferencesMap: ImmutableDictionary<ProjectId, ImmutableHashSet<ProjectId>>.Empty,
+                  default,
+                  default)
         {
+        }
+
+        // This constructor is private to prevent other Roslyn code from producing this type with inconsistent inputs.
+        private ProjectDependencyGraph(
+            ImmutableHashSet<ProjectId> projectIds,
+            ImmutableDictionary<ProjectId, ImmutableHashSet<ProjectId>> referencesMap,
+            ImmutableDictionary<ProjectId, ImmutableHashSet<ProjectId>> reverseReferencesMap,
+            ImmutableDictionary<ProjectId, ImmutableHashSet<ProjectId>> transitiveReferencesMap,
+            ImmutableDictionary<ProjectId, ImmutableHashSet<ProjectId>> reverseTransitiveReferencesMap,
+            ImmutableArray<ProjectId> topologicallySortedProjects,
+            ImmutableArray<IEnumerable<ProjectId>> dependencySets)
+        {
+            Contract.ThrowIfNull(transitiveReferencesMap);
+            Contract.ThrowIfNull(reverseTransitiveReferencesMap);
+
             _projectIds = projectIds;
             _referencesMap = referencesMap;
+            _lazyReverseReferencesMap = reverseReferencesMap;
+            _transitiveReferencesMap = transitiveReferencesMap;
+            _reverseTransitiveReferencesMap = reverseTransitiveReferencesMap;
+            _lazyTopologicallySortedProjects = topologicallySortedProjects;
+            _lazyDependencySets = dependencySets;
+        }
+
+        internal ProjectDependencyGraph WithAdditionalProjects(IEnumerable<ProjectId> projectIds)
+        {
+            // Track the existence of some new projects. Note this call only adds new ProjectIds, but doesn't add any references. Any caller who wants to add a new project
+            // with references will first call this, and then call WithAdditionalProjectReferences to add references as well.
+            var newTopologicallySortedProjects = _lazyTopologicallySortedProjects;
+
+            if (!newTopologicallySortedProjects.IsDefault)
+            {
+                newTopologicallySortedProjects = newTopologicallySortedProjects.AddRange(projectIds);
+            }
+
+            var newDependencySets = _lazyDependencySets;
+
+            if (!newDependencySets.IsDefault)
+            {
+                var builder = newDependencySets.ToBuilder();
+
+                foreach (var projectId in projectIds)
+                {
+                    builder.Add(ImmutableArray.Create(projectId));
+                }
+
+                newDependencySets = builder.ToImmutable();
+            }
+
+            return new ProjectDependencyGraph(
+                _projectIds.Union(projectIds),
+                referencesMap: _referencesMap,
+                reverseReferencesMap: _lazyReverseReferencesMap,
+                transitiveReferencesMap: _transitiveReferencesMap,
+                reverseTransitiveReferencesMap: _reverseTransitiveReferencesMap,
```

Comment:
it's really worth explaining why certain fields are not updated (even though the fields also explain it).  But it would still be helpful.  Both for the cases where we don't actually need to update them because they haven't changed, or we don't need to update them because we'll compute their correct value the next time they're asked.